langroid/agent/task.py

```diff
@@ -895,8 +925,37 @@ async def run_async(
         if return_type is None:
             return_type = self.default_return_type
 
+        # If possible, take a final strict decoding step
+        # when the output does not match `return_type`
         if return_type is not None and return_type != ChatDocument:
-            return self.agent.from_ChatDocument(final_result, return_type)
+            parsed_result = self.agent.from_ChatDocument(final_result, return_type)
```

Comment:
the async is an exact dupe of the sync, so maybe use a helper function to encapsulate this code block?